components/remote_settings/src/client.rs

```diff
@@ -0,0 +1,551 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::config::RemoteSettingsConfig;
+use crate::error::{RemoteSettingsError, Result};
+use crate::UniffiCustomTypeConverter;
+use parking_lot::Mutex;
+use serde::Deserialize;
+use std::time::{Duration, Instant};
+use url::Url;
+use viaduct::{status_codes, Request, Response};
+
+const HEADER_BACKOFF: &str = "Backoff";
+const HEADER_RETRY_AFTER: &str = "Retry-After";
+
+/// A simple HTTP client that can retrieve Remote Settings data using the properties by [ClientConfig].
+/// Methods defined on this will fetch data from
+/// <base_url>/v1/buckets/<bucket_name>/collections/<collection_name>/
+pub struct Client {
+    pub(crate) base_url: Url,
+    pub(crate) bucket_name: String,
+    pub(crate) collection_name: String,
+    pub(crate) remote_state: Mutex<RemoteState>,
+}
+
+impl Client {
+    /// Create a new [Client] with properties matching config.
+    /// During construction, the [Client] will fetch the base attachments
+    /// URL from the server's metadata, assuming attachments are enabled for
+    /// the server.
+    pub fn new(config: RemoteSettingsConfig) -> Result<Self> {
+        let server_url = config
+            .server_url
+            .unwrap_or_else(|| String::from("https://firefox.settings.services.mozilla.com"));
+        let bucket_name = config.bucket_name.unwrap_or_else(|| String::from("main"));
+        let base_url = Url::parse(&server_url)?;
+
+        Ok(Self {
+            base_url,
+            bucket_name,
+            collection_name: config.collection_name,
+            remote_state: Default::default(),
+        })
+    }
+
+    /// Fetches all records for a collection that can be found in the server,
+    /// bucket, and collection defined by the [ClientConfig] used to generate
+    /// this [Client].
+    pub fn get_records(&self) -> Result<RemoteSettingsResponse> {
+        let path = format!(
+            "v1/buckets/{}/collections/{}/records",
+            &self.bucket_name, &self.collection_name
+        );
+        let resp = self.make_request(path)?;
+        let records = resp.json::<RecordsResponse>()?.data;
```

Comment:
Will this work correctly for a 304? If the body was empty (which I think is typical with that response?) then I suspect this json parsing might fail.

Reply:
Oh, nice catch! I wonder if we can remove the 304 check entirely, since we aren't (AFAICT?) ever sending an `If-None-Match` or an `If-Modified-Since` request header, and Kinto's docs (and RFC 9110) say that's the only time we'll receive a 304 response.

Reply:
ah - I went looking for them too, but then noticed the "since" param in the url, which I thought might trigger it - but yeah, you are probably correct that in that case we'll get a 200 with no records - so yeah, if that's the case, I agree we should kill the 304 handling.

Reply:
Just to confirm, the above all means that we don't want to add explicit 304 handling?

Reply:
> the above all means that we don't want to add explicit 304 handling?

Correct.

I double-checked, and `curl "https://firefox.settings.services.mozilla.com/v1/buckets/main/collections/quicksuggest/records?_since=1683313881574"` (where `1683313881574` is an `ETag` that I got from a previous request without the `_since`), indeed, returns a 200 with no records:

```
< HTTP/2 200
< ...
{"data":[]}
```

...So we don't need the explicit 304 handling! 🚀